File: day08/part1/main.dart
// -*- compile-command: "dart main.dart ../example1.txt ../input.txt"; -*-

import 'dart:io';

main(List<String> args) {
  for (var arg in args) {
    process(arg);
  }

  print('Done.');
}

class Instruction {
  String op;
  int arg;
  Instruction({this.op, this.arg});

  toString() => 'op: $op, arg: $arg';
}

process(String filename) {
  print('Processing $filename ...');

  var program = List<Instruction>();
  var lines = File(filename).readAsLinesSync();
  for (var line in lines) {
    var parts = line.split(' ');
    var ins = Instruction(op: parts[0], arg: int.parse(parts[1]));
    program.add(ins);
  }

  var cpu = CPU(program: program);
  cpu.execute();
}

class CPU {
  int accumulator;
  int ip;
  List<Instruction> program;
  CPU({
      this.accumulator = 0,
      this.ip = 0,
      this.program,
  });

  execute() {
    var seen = Set<int>();

    while (true) {
      if (seen.contains(ip)) {
        print('Solution Accumulator = $accumulator');
        return;
      }
      seen.add(ip);

      var ins = program[ip];
      switch (ins.op) {
        case 'nop':
        ip++;
        break;
        case 'acc':
        accumulator += ins.arg;
        ip++;
        break;
        case 'jmp':
        ip += ins.arg;
        break;
      }
    }
  }
}

File: day08/part1/main.go
// -*- compile-command: "go run main.go ../example1.txt ../input.txt"; -*-

package main

import (
	"flag"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
)

var (
	verbose = flag.Bool("v", false, "Verbose log messages")
)

func main() {
	flag.Parse()

	for _, arg := range flag.Args() {
		process(arg)
	}

	log.Printf("Done.")
}

type CPU struct {
	accumulator int
	ip          int
	program     []*Instruction
}

type Instruction struct {
	op  string
	arg int
}

func process(filename string) {
	cpu := &CPU{}

	lines := readLines(filename)
	// log.Printf("lines=%v", lines)
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		parts := strings.Split(line, " ")
		op := parts[0]
		arg, err := strconv.Atoi(parts[1])
		check("strconv.Atoi: %v", err)
		cpu.program = append(cpu.program, &Instruction{op: op, arg: arg})
	}

	cpu.execute()

	// fmt.Printf("Solution: %v\n", count-1)
}

func (cpu *CPU) execute() {
	seen := map[int]bool{}

	for {
		if seen[cpu.ip] {
			log.Printf("Solution Accumulator = %v", cpu.accumulator)
			break
		}
		seen[cpu.ip] = true

		arg := cpu.program[cpu.ip].arg
		switch cpu.program[cpu.ip].op {
		case "nop":
			cpu.ip++
		case "acc":
			cpu.accumulator += arg
			cpu.ip++
		case "jmp":
			cpu.ip += arg
		}
	}
}

func readLines(filename string) []string {
	log.Printf("Processing %v ...", filename)
	buf, err := ioutil.ReadFile(filename)
	check("ReadFile: %v", err)
	lines := strings.Split(string(buf), "\n")
	return lines
}

func check(fmtStr string, args ...interface{}) {
	if err := args[len(args)-1]; err != nil {
		log.Fatalf(fmtStr, args...)
	}
}

func logf(fmt string, args ...interface{}) {
	if *verbose {
		log.Printf(fmt, args...)
	}
}
